### PR TITLE
PCC02 LaneHesser

### PR DESCRIPTION
**Pull request by LaneHesser**
Difficulty level (1-10): [4]
Estimated time spent (hours): [4]
Completed (yes/no): [Yes]
I stretched my coding skills (if yes what did you learn?): [I learned some useful functions from the itertools module and how to put a few of them together to generate all permutations of the items in a sequence.]
Other feedback (what can we improve?): []

### CODE REVIEW

**Review by LaneHesser — COMMENTED** [review]
Realized I had left out some functionality when I pushed the first time, so I made the changes and pushed again.

File: 02/LaneHesser/game.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    return [
+        random.choice(POUCH)
+        for _ in range(NUM_LETTERS)
+    ]
+
+
+def get_user_input(draw):
+    """Asks the player to form a word with one or more letters from draw."""
+    draw_copy = draw.copy()
+    print('Letters drawn:', *draw_copy)
```
Comment (bbelderbos) [inline 02/LaneHesser/game.py:58]: f-string?

File: 02/LaneHesser/game.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+def _permutations(iterable):
+    """Returns all possible permutations of the letters in draw."""
+    return chain.from_iterable(permutations(iterable, n)
+                               for n in range(1, NUM_LETTERS+1))
+
+
+def calc_optimal_word(draw):
+    """Calculates optimal word (highest value) given the letters in draw"""
+    permutations = _permutations(draw)
```
Comment (bbelderbos) [inline 02/LaneHesser/game.py:34]: I would call this differently, it clashes with `itertools.permutations`

File: 02/LaneHesser/game.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    valid_words = set(
+        ''.join(word).lower()
+        for word in permutations
+        if ''.join(word).lower() in DICTIONARY
```
Comment (bbelderbos) [inline 02/LaneHesser/game.py:39]: wonder if you can do the `lower()` in `_permutations` to not do it twice here?

**Review by bbelderbos — APPROVED** [review]
Great work, some feedback

**Comment by LaneHesser** [comment]
Awesome. Thank you for the feedback. It is much appreciated.

Lane Hesser

**Pull request merged**